File: backend/app/models/evaluator.py
"""
Evaluator models
"""
from sqlalchemy import Column, Integer, String, Text, DateTime, JSON, Enum as SQLEnum, ForeignKey, Boolean
from sqlalchemy.orm import relationship
from datetime import datetime
from app.core.database import Base
import enum


class EvaluatorType(str, enum.Enum):
    PROMPT = "prompt"
    CODE = "code"


class EvaluatorBoxType(str, enum.Enum):
    WHITE = "white"  # 白盒
    BLACK = "black"  # 黑盒


class EvaluatorVersionStatus(str, enum.Enum):
    DRAFT = "draft"
    SUBMITTED = "submitted"
    ARCHIVED = "archived"


class Evaluator(Base):
    __tablename__ = "evaluators"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, index=True)
    description = Column(Text, nullable=True)
    evaluator_type = Column(SQLEnum(EvaluatorType, values_callable=lambda x: [e.value for e in x]), nullable=False)
    latest_version = Column(String(50), nullable=True)
    
    # New fields
    builtin = Column(Boolean, default=False, nullable=False)  # 是否内置评估器
    box_type = Column(SQLEnum(EvaluatorBoxType, values_callable=lambda x: [e.value for e in x]), nullable=True)  # 黑白盒类型
    evaluator_info = Column(JSON, nullable=True)  # benchmark, vendor, vendor_url, user_manual_url
    tags = Column(JSON, nullable=True)  # 标签系统，格式: {"zh-CN": {"Category": [...], ...}, "en-US": {...}}
    
    # Content fields (directly stored in evaluator, synced with versions for backward compatibility)
    prompt_content = Column(JSON, nullable=True)  # For Prompt evaluators: message_list, model_config, etc.
    code_content = Column(JSON, nullable=True)  # For Code evaluators: code_content, language_type, etc.
    input_schemas = Column(JSON, nullable=True)  # 输入 Schema 定义
    output_schemas = Column(JSON, nullable=True)  # 输出 Schema 定义
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(String(100), nullable=True)

    # Relationships
    versions = relationship("EvaluatorVersion", back_populates="evaluator", cascade="all, delete-orphan")


class EvaluatorVersion(Base):
    __tablename__ = "evaluator_versions"

    id = Column(Integer, primary_key=True, index=True)
    evaluator_id = Column(Integer, ForeignKey("evaluators.id"), nullable=False)
    version = Column(String(50), nullable=False)
    description = Column(Text, nullable=True)
    status = Column(SQLEnum(EvaluatorVersionStatus, values_callable=lambda x: [e.value for e in x]), default=EvaluatorVersionStatus.DRAFT, nullable=False)
    
    # Schema definitions
    input_schemas = Column(JSON, nullable=True)  # 输入 Schema 定义
    output_schemas = Column(JSON, nullable=True)  # 输出 Schema 定义
    
    # Type-specific content
    # For Prompt: message_list, model_config, tools, parse_type, prompt_suffix, receive_chat_history
    prompt_content = Column(JSON, nullable=True)
    # For Code: code_content, language_type, code_template_key, code_template_name
    code_content = Column(JSON, nullable=True)
    
    # Legacy content field (for backward compatibility)
    content = Column(JSON, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(String(100), nullable=True)

    # Relationships
    evaluator = relationship("Evaluator", back_populates="versions")
    records = relationship("EvaluatorRecord", back_populates="evaluator_version", cascade="all, delete-orphan")

File: frontend/src/types/evaluator.ts
/**
 * Evaluator type definitions
 */

export type EvaluatorType = 'prompt' | 'code'
export type EvaluatorBoxType = 'white' | 'black'
export type EvaluatorVersionStatus = 'draft' | 'submitted' | 'archived'
export type EvaluatorRunStatus = 'unknown' | 'success' | 'fail'
export type ContentType = 'Text' | 'Image' | 'Audio' | 'MultiPart'
export type ParseType = 'json' | 'text'
export type LanguageType = 'Python' | 'JS'
export type Role = 'system' | 'user' | 'assistant' | 'tool'

export interface EvaluatorInfo {
  benchmark?: string
  vendor?: string
  vendor_url?: string
  user_manual_url?: string
}

export interface EvaluatorTags {
  'zh-CN'?: Record<string, string[]>
  'en-US'?: Record<string, string[]>
}

export interface Evaluator {
  id: number
  name: string
  description?: string
  evaluator_type: EvaluatorType
  latest_version?: string
  builtin?: boolean
  box_type?: EvaluatorBoxType
  evaluator_info?: EvaluatorInfo
  tags?: EvaluatorTags
  // Content fields (directly stored in evaluator)
  prompt_content?: PromptEvaluatorContent
  code_content?: CodeEvaluatorContent
  input_schemas?: ArgsSchema[]
  output_schemas?: ArgsSchema[]
  created_at?: string
  updated_at?: string
  created_by?: string
}

export interface ArgsSchema {
  key?: string
  support_content_types?: ContentType[]
  json_schema?: string
  default_value?: Content
}

export interface Content {
  content_type?: ContentType
  format?: string
  text?: string
  image?: {
    name?: string
    url?: string
    uri?: string
    thumb_url?: string
  }
  audio?: {
    format?: string
    url?: string
  }
  multi_part?: Content[]
}

export interface Message {
  role: Role
  content?: Content
  ext?: Record<string, string>
}

export interface ModelConfig {
  provider?: string
  model: string
  api_key?: string
  api_base?: string
  temperature?: number
  max_tokens?: number
  top_p?: number
  frequency_penalty?: number
  presence_penalty?: number
  timeout?: number
  extra_config?: Record<string, any>
  // New field: model configuration ID from model_configs table
  model_config_id?: number
}

export interface PromptEvaluatorContent {
  message_list: Message[]
  model_config: ModelConfig
  tools?: any[]
  parse_type?: ParseType
  prompt_suffix?: string
  receive_chat_history?: boolean
}

export interface CodeEvaluatorContent {
  code_content: string
  language_type: LanguageType
  code_template_key?: string
  code_template_name?: string
}

export interface EvaluatorVersion {
  id: number
  evaluator_id: number
  version: string
  description?: string
  status: EvaluatorVersionStatus
  input_schemas?: ArgsSchema[]
  output_schemas?: ArgsSchema[]
  prompt_content?: PromptEvaluatorContent
  code_content?: CodeEvaluatorContent
  content?: Record<string, any> // Legacy field
  created_at?: string
  updated_at?: string
  created_by?: string
}

export interface EvaluatorInputData {
  history_messages?: Message[]
  input_fields?: Record<string, Content>
  evaluate_dataset_fields?: Record<string, Content>
  evaluate_target_output_fields?: Record<string, Content>
  ext?: Record<string, string>
}

export interface EvaluatorResult {
  score?: number
  correction?: {
    score?: number
    explain?: string
    updated_by?: string
  }
  reasoning?: string
}

export interface EvaluatorUsage {
  input_tokens?: number
  output_tokens?: number
}

export interface EvaluatorRunError {
  code?: number
  message?: string
}

export interface EvaluatorOutputData {
  evaluator_result?: EvaluatorResult
  evaluator_usage?: EvaluatorUsage
  evaluator_run_error?: EvaluatorRunError
  time_consuming_ms?: number
  stdout?: string
}

export interface EvaluatorRecord {
  id: number
  evaluator_version_id: number
  experiment_id?: number
  experiment_run_id?: number
  dataset_item_id?: number
  turn_id?: number
  input_data: EvaluatorInputData
  output_data: EvaluatorOutputData
  status: EvaluatorRunStatus
  trace_id?: string
  log_id?: string
  ext?: Record<string, string>
  created_at?: string
  created_by?: string
}

